File: Structural/Lab10/C#/Entities/Agency.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_.Entities
{
    public class Agency
    {
        private Agency() {
            _apartments = new List<Apartment>();
            _realtors = new List<Realtor>();
            _houses = new List<House>();
            _rooms = new List<Room>();
        }

        //Instances
        private static Agency _instance;
        private static List<Apartment> _apartments;
        private static List<Room> _rooms;
        private static List<House> _houses;
        private static List<Realtor> _realtors;

        //Locks
        private static readonly object _lockInstance = new object();
        private static readonly object _lockApartments = new object();
        private static readonly object _lockRooms = new object();
        private static readonly object _lockHouses = new object();
        private static readonly object _lockRealtors = new object();

        public static Agency GetInstance()
        {
            if (_instance == null)
            {
                lock (_lockInstance)
                {
                    if (_instance == null)
                        _instance = new Agency();
                }
            }
            return _instance;
        }

        public  List<Apartment> Apartments 
        {
            get 
            {
                lock (_lockApartments)
                {
                    return _apartments;
                }
            }
        }
        public List<Room> Rooms
        {
            get 
            {
                lock (_lockRooms)
                {
                    return _rooms;
                }
            }
        }
        public List <House> Houses
        {
            get 
            {
                lock (_lockHouses)
                {
                    return _houses;
                }
            }
        }
        public List<Realtor> Realtors 
        { 
            get 
            {
                lock (_lockRealtors)
                {
                    return _realtors;
                }
            }
        }

        public void AddRealtor(Realtor realtor){
            lock (_lockRealtors) {
                _realtors.Add(realtor);
            }
        }
        public void RemoveRealtor(Realtor realtor){
            lock(_lockRealtors) {
                _realtors.Remove(realtor);
            }
        }

        public void AddEstate(RealEstate estate){
            switch (estate.GetType().Name) {
                case "Apartment":
                    lock(_lockApartments){
                        _apartments.Add((Apartment)estate);
                    };
                    break;
                case "Room":
                    lock(_lockRooms){
                        _rooms.Add((Room)estate);
                    };
                    break;
                case "House":
                    lock(_lockHouses){
                        _houses.Add((House)estate);
                    };
                    break;
            }
        }

        public void RemoveEstate(RealEstate estate){
            switch (estate.GetType().Name) {
                case "Apartment":
                    lock(_lockApartments){
                        _apartments.Remove((Apartment)estate);
                    };
                    break;
                case "Room":
                    lock(_lockRooms){
                        _rooms.Remove((Room)estate);
                    };
                    break;
                case "House":
                    lock(_lockHouses){
                        _houses.Remove((House)estate);
                    };
                    break;
            }
        }

        
    }
}

File: Structural/Lab13/Typescript/src/Entities/agency.ts
import { Apartment } from './apartment';
import { Room } from './room';
import { House } from './house';
import { Realtor } from './realtor';
import { RealEstate } from './real_estate';
import { IAgency } from '../Interfaces/agency';
export class Agency implements IAgency {
    private static _instance: Agency;
    private static _apartments: Apartment[];
    private static _rooms: Room[];
    private static _houses: House[];
    private static _realtors: Realtor[];

    private static readonly _lockInstance = new Object();
    private static readonly _lockApartments = new Object();
    private static readonly _lockRooms = new Object();
    private static readonly _lockHouses = new Object();
    private static readonly _lockRealtors = new Object();

    private constructor() {
        Agency._apartments = [];
        Agency._realtors = [];
        Agency._houses = [];
        Agency._rooms = [];
    }

    public static GetInstance(): Agency {
        if (!Agency._instance) {
                if (!Agency._instance) {
                    Agency._instance = new Agency();
                }
        }
        return Agency._instance;
    }

    public get Apartments(): Apartment[] {
        return Agency._apartments;
    }

    public get Rooms(): Room[] {
            return Agency._rooms;
        
    }

    public get Houses(): House[] {
            return Agency._houses;
    }

    public get Realtors(): Realtor[] {
            return Agency._realtors;
        
    }

    public AddRealtor(realtor: Realtor): void {
            Agency._realtors.push(realtor);
        
    }

    public RemoveRealtor(realtor: Realtor): void {
        const index = Agency._realtors.indexOf(realtor);
        if (index > -1) {
            Agency._realtors.splice(index, 1);
        }
    }
    

    public AddEstate(estate: RealEstate): void {
        switch (estate.constructor.name) {
            case "Apartment":
                Agency._apartments.push(estate as Apartment);
                break;
            case "Room":
                Agency._rooms.push(estate as Room);
                break;
            case "House":
                Agency._houses.push(estate as House);
                break;
        }
    }

    public RemoveEstate(estate: RealEstate): void {
        switch (estate.constructor.name) {
            case "Apartment":
                if (Agency._apartments.indexOf(estate as Apartment) > -1) {
                    Agency._apartments.splice(Agency._apartments.indexOf(estate as Apartment), 1);
                }
                break;
            case "Room":
                if (Agency._rooms.indexOf(estate as Room) > -1) {
                    Agency._rooms.splice(Agency._rooms.indexOf(estate as Room), 1);
                }
                break;
            case "House":
                if (Agency._houses.indexOf(estate as House) > -1) {
                    Agency._houses.splice(Agency._houses.indexOf(estate as House), 1);
                }
                break;
        }
    }

    public FindEstateByID(id: number): RealEstate {
        return Agency._apartments.find(a => a.id === id) as RealEstate ||
            Agency._rooms.find(a => a.id === id) as RealEstate ||
            Agency._houses.find(a => a.id === id);
    }
}